metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -1939,40 +1938,40 @@ class MetalsLanguageServer(
       case other => Future.successful(other)
     }
 
-    {
-      for {
-        _ <- disconnectOldBuildServer()
-        maybeSession <- timerProvider.timed("connected to build server", true) {
-          bspConnector.connect(workspace, userConfig)
-        }
-        result <- maybeSession match {
-          case Some(session) =>
-            val result = connectToNewBuildServer(session)
-            session.mainConnection.onReconnection { newMainConn =>
-              val updSession = session.copy(main = newMainConn)
-              connectToNewBuildServer(updSession)
-                .flatMap(compileAllOpenFiles)
-                .ignoreValue
-            }
-            result
-          case None =>
-            Future.successful(BuildChange.None)
-        }
-        _ = {
-          treeView.init()
-        }
-      } yield result
-    }.recover { case NonFatal(e) =>
-      disconnectOldBuildServer()
-      val message =
-        "Failed to connect with build server, no functionality will work."
-      val details = " See logs for more details."
-      languageClient.showMessage(
-        new MessageParams(MessageType.Error, message + details)
-      )
-      scribe.error(message, e)
-      BuildChange.Failed
-    }.flatMap(compileAllOpenFiles)
+    (for {
+      _ <- disconnectOldBuildServer()
+      maybeSession <- timerProvider.timed("Connected to build server", true) {
+        bspConnector.connect(workspace, userConfig)
+      }
+      result <- maybeSession match {
+        case Some(session) =>
+          val result = connectToNewBuildServer(session)
+          session.mainConnection.onReconnection { newMainConn =>
+            val updSession = session.copy(main = newMainConn)
+            connectToNewBuildServer(updSession)
+              .flatMap(compileAllOpenFiles)
+              .ignoreValue
+          }
+          result
+        case None =>
+          Future.successful(BuildChange.None)
+      }
+      _ = {
+        treeView.init()
+      }
+    } yield result)
+      .recover { case NonFatal(e) =>
+        disconnectOldBuildServer()
+        val message =
+          "Failed to connect with build server, no functionality will work."
+        val details = " See logs for more details."
+        languageClient.showMessage(
+          new MessageParams(MessageType.Error, message + details)
+        )
+        scribe.error(message, e)
+        BuildChange.Failed
+      }
+      .flatMap(compileAllOpenFiles)
```

Comment:
I hesitated to completely redo this, but wanted to instead just focus on the task at hand. However, I think we can actually improve this here. Mainly, in the first part we are actually disconnecting and then re-connecting. Then we are running the full build import, and then initiating the tree view. However, if at any point during that process something goes wrong, we `disconnectOldBuildServer()` when that's probably not necessary at all depending on the situation. It may be preferable to instead break those three things out from one another and handle a failure in differently. An argument could be made that if something goes wonky during the build import that _maybe_ could be fixed easily, do we really want to disconnect fully? There may be partial support still very much available. Just something to think about.

Reply:
We should not fail on tree view for sure, we used to fail on compile too, but I moved it out because of that. It's fine to do it here, but also can be a separate PR.